compiler/rustc_lint/src/unnecessary_send_constraint.rs

```diff
@@ -0,0 +1,56 @@
+use rustc_span::sym;
+
+use crate::hir;
+
+use crate::{lints::UnnecessarySendConstraintDiag, LateContext, LateLintPass};
+
+declare_lint! {
+    /// The `lint_unnecessary_send_constraint` lints unnecessary constraint of references to `Send`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// fn foo(_: &(dyn Any + Send>) {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// References cannot be sent across threads unless they have a `Sync` bound, so constraining them to `Send` without `Sync` is unnecessary.
+    pub UNNECESSARY_SEND_CONSTRAINT,
+    Warn,
+    "constraining a reference to `Send` without `Sync` is unnecessary, consider removing it"
```

Comment:
Shouldn't the recommendation be replace `Send` by `Sync`? That seems more likely to be what the author intended.

Reply:
See this comment: https://github.com/rust-lang/rust/pull/110961#issuecomment-1550271436

In those situations, it stems from a `Box<dyn Any + Send>` to which a &dyn was made. It's easy to just copy `Any + Send`, but the Send part should be removed for shared references.

Reply:
That seems like a fairly special case to me, I would not expect this to be common enough to warrant a general recommendation of dropping the `Send`. The more symmetric thing to do would be to have `&dyn Any+Sync`, to get a shared reference that can still be shared across thread boundaries. Of course then you'd need `Box<dyn Any+Send+Sync>` for the owned case.

Reply:
So it should recommend replacing `Send` with `Sync` instead? just confirming before making the change